pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -675,32 +674,69 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
 	if m.allocatedDevices[resource] == nil {
 		m.allocatedDevices[resource] = sets.NewString()
 	}
+
 	// Gets Devices in use.
 	devicesInUse := m.allocatedDevices[resource]
-	// Gets a list of available devices.
+	// Gets Available devices.
 	available := m.healthyDevices[resource].Difference(devicesInUse)
 	if available.Len() < needed {
 		return nil, fmt.Errorf("requested number of devices unavailable for %s. Requested: %d, Available: %d", resource, needed, available.Len())
 	}
-	// By default, pull devices from the unsorted list of available devices.
-	allocated := available.UnsortedList()[:needed]
-	// If topology alignment is desired, update allocated to the set of devices
-	// with the best alignment.
-	hint := m.topologyAffinityStore.GetAffinity(podUID, contName)
-	if m.deviceHasTopologyAlignment(resource) && hint.NUMANodeAffinity != nil {
-		allocated = m.takeByTopology(resource, available, hint.NUMANodeAffinity, needed)
+
+	// Filters available Devices based on NUMA affinity.
+	aligned, unaligned, noAffinity := m.filterByAffinity(podUID, contName, resource, available)
+
+	// Allocates from aligned Devices list.
+	for device := range aligned.Difference(devices) {
+		m.allocatedDevices[resource].Insert(device)
+		devices.Insert(device)
+		needed--
+		if needed == 0 {
+			return devices, nil
+		}
+	}
+	// Calls GetPreferredAllocation if it is available.
+	preferred, err := m.callGetPreferredAllocationIfAvailable(podUID, contName, resource, available.Union(reusableDevices), devices, required)
+	if err != nil {
+		return nil, err
+	}
+	// Allocates from the preferred Devices list.
+	for device := range preferred.Difference(devices) {
+		m.allocatedDevices[resource].Insert(device)
+		devices.Insert(device)
+		needed--
+		if needed == 0 {
+			return devices, nil
+		}
+	}
```

Comment:
I would have thought allocation from the preferred list would take place before the aligned list. Is this because we can't assume preferred devices are aligned also?

Reply:
Exactly. The goal is to first honour allocations based on topologymanager alignment, and only then defer to the plugin to help decide how to alllocate the rest.

Reply:
Updated with detailed comments in latest